src/box/sql/build.c

```diff
@@ -1194,85 +1196,28 @@ vdbe_emit_fk_constraint_create(struct Parse *parse_context,
-				  constr_tuple_reg + 2);
+	if (!is_alter_add_constr)
+		sqlVdbeAddOp2(vdbe, OP_SCopy, fk->child_id, regs);
+	else
+		sqlVdbeAddOp2(vdbe, OP_Integer, fk->child_id, regs);
+	if (!is_alter_add_constr && fk_constraint_is_self_referenced(fk))
+		sqlVdbeAddOp2(vdbe, OP_SCopy, fk->parent_id, regs + 1);
+	else
+		sqlVdbeAddOp2(vdbe, OP_Integer, fk->parent_id, regs + 1);
+	if (fk->is_field_fk) {
+		sqlVdbeAddOp2(vdbe, OP_Integer, fk->links->child_field,
+			      regs + 2);
+		sqlVdbeAddOp2(vdbe, OP_Integer, fk->links->parent_field,
+			      regs + 3);
 	} else {
-		sqlVdbeAddOp2(vdbe, OP_Integer, fk->parent_id,
-				  constr_tuple_reg + 2);
+		uint32_t size;
+		char *raw = fk_constraint_encode_links(fk, &size);
+		sqlVdbeAddOp4(vdbe, OP_Blob, size, regs + 2,
+			      SQL_SUBTYPE_MSGPACK, raw, P4_DYNAMIC);
 	}
-	/*
-	 * Lets check that constraint with this name hasn't
-	 * been created before.
-	 */
-	const char *error_msg =
-		tt_sprintf(tnt_errcode_desc(ER_CONSTRAINT_EXISTS),
-			   constraint_type_strs[CONSTRAINT_TYPE_FK], name_copy,
-			   space_name);
-	if (vdbe_emit_halt_with_presence_test(parse_context,
-					      BOX_FK_CONSTRAINT_ID, 0,
-					      constr_tuple_reg, 2,
-					      ER_CONSTRAINT_EXISTS, error_msg,
-					      false, OP_NoConflict) != 0)
-		return;
-	sqlVdbeAddOp2(vdbe, OP_Bool, false, constr_tuple_reg + 3);
-	sqlVdbeAddOp4(vdbe, OP_String8, 0, constr_tuple_reg + 4, 0, "simple",
-		      P4_STATIC);
-	sqlVdbeAddOp4(vdbe, OP_String8, 0, constr_tuple_reg + 5, 0, "no_action",
-		      P4_STATIC);
-	sqlVdbeAddOp4(vdbe, OP_String8, 0, constr_tuple_reg + 6, 0, "no_action",
-		      P4_STATIC);
-	struct region *region = &parse_context->region;
-	uint32_t parent_links_size = 0;
-	char *parent_links = fk_constraint_encode_links(region, fk, FIELD_LINK_PARENT,
-					       &parent_links_size);
-	if (parent_links == NULL)
-		goto error;
-	uint32_t child_links_size = 0;
-	char *child_links = fk_constraint_encode_links(region, fk, FIELD_LINK_CHILD,
-					      &child_links_size);
-	if (child_links == NULL)
-		goto error;
-	/*
-	 * We are allocating memory for both parent and child
-	 * arrays in the same chunk. Thus, first OP_Blob opcode
-	 * interprets it as static memory, and the second one -
-	 * as dynamic and releases memory.
-	 */
-	char *raw = sqlDbMallocRaw(parse_context->db,
-				       parent_links_size + child_links_size);
-	if (raw == NULL)
-		return;
-	memcpy(raw, parent_links, parent_links_size);
-	parent_links = raw;
-	raw += parent_links_size;
-	memcpy(raw, child_links, child_links_size);
-	child_links = raw;
-
-	sqlVdbeAddOp4(vdbe, OP_Blob, child_links_size, constr_tuple_reg + 7,
-			  SQL_SUBTYPE_MSGPACK, child_links, P4_STATIC);
-	sqlVdbeAddOp4(vdbe, OP_Blob, parent_links_size,
-			  constr_tuple_reg + 8, SQL_SUBTYPE_MSGPACK,
-			  parent_links, P4_DYNAMIC);
-	sqlVdbeAddOp3(vdbe, OP_MakeRecord, constr_tuple_reg, 9,
-			  constr_tuple_reg + 9);
-	sqlVdbeAddOp2(vdbe, OP_SInsert, BOX_FK_CONSTRAINT_ID,
```

Comment:
Don't remember if I have already asked that somewhere. Is there a need to update upgrade.lua script to port existing FKs to the new storage way? Did you test, maybe manually, that old FKs stored in the space `BOX_FK_CONSTRAINT_ID` still work and their deletion also works?

Reply:
There is a patch that updates all existing foreign key and check constraints in the PR about check constraint. I have verified that the SQL FK still works, and even after the update patch it will be possible to create the SQL FK by inserting its definition into `_fk_coinstraint`. Not for long, as I plan to completely drop all SQL FK/CK code, with the exception of some structures that are still in use.